From 809b5e8657066133f9208fa92cf075a1c9e91aad Mon Sep 17 00:00:00 2001
From: Koji Matsuoka <koji.matsuoka.xm@renesas.com>
Date: Thu, 11 Oct 2018 12:53:07 +0900
Subject: [PATCH 359/909] media: rcar-vin: Fix VnCSI_IFMD register access for
 r8a77990

commit 2a626e26cb0b0e30ba88d0d122d2452b412fbc4d from
git://git.kernel.org/pub/scm/linux/kernel/git/horms/renesas-bsp.git

According to H/W manual v0.54, DES1 bit in VnCSI_IFMD register
is reserved in r8a77990. This patch corresponds it.

Signed-off-by: Koji Matsuoka <koji.matsuoka.xm@renesas.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/media/platform/rcar-vin/rcar-core.c | 14 +++++++++++++-
 drivers/media/platform/rcar-vin/rcar-dma.c  |  3 +++
 drivers/media/platform/rcar-vin/rcar-vin.h  |  4 ++++
 3 files changed, 20 insertions(+), 1 deletion(-)

diff --git a/drivers/media/platform/rcar-vin/rcar-core.c b/drivers/media/platform/rcar-vin/rcar-core.c
index 84e2ef6c0c85..6b387acaa5bb 100644
--- a/drivers/media/platform/rcar-vin/rcar-core.c
+++ b/drivers/media/platform/rcar-vin/rcar-core.c
@@ -1134,9 +1134,17 @@ static const struct soc_device_attribute r8a7795es1[] = {
 	{ /* Sentinel */ }
 };
 
+static const struct soc_device_attribute chip_info[] = {
+	{
+		.soc_id = "r8a77990",
+		.data = (void *)RCAR_VIN_DES1_RESERVED,
+	},
+	{ /* sentinel */ }
+};
+
 static int rcar_vin_probe(struct platform_device *pdev)
 {
-	const struct soc_device_attribute *attr;
+	const struct soc_device_attribute *attr, *dev_attr;
 	struct rvin_dev *vin;
 	struct resource *mem;
 	int irq, ret;
@@ -1156,6 +1164,10 @@ static int rcar_vin_probe(struct platform_device *pdev)
 	if (attr)
 		vin->info = attr->data;
 
+	dev_attr = soc_device_match(chip_info);
+	if (dev_attr)
+		vin->chip_info = (uintptr_t)dev_attr->data;
+
 	mem = platform_get_resource(pdev, IORESOURCE_MEM, 0);
 	if (mem == NULL)
 		return -EINVAL;
diff --git a/drivers/media/platform/rcar-vin/rcar-dma.c b/drivers/media/platform/rcar-vin/rcar-dma.c
index 0e50dc21f4ff..2fb048ed6ba6 100644
--- a/drivers/media/platform/rcar-vin/rcar-dma.c
+++ b/drivers/media/platform/rcar-vin/rcar-dma.c
@@ -1754,6 +1754,9 @@ int rvin_set_channel_routing(struct rvin_dev *vin, u8 chsel)
 
 	vin->chsel = chsel;
 
+	if (vin->chip_info & RCAR_VIN_DES1_RESERVED)
+		ifmd &= (u32)~VNCSI_IFMD_DES1;
+
 	rvin_write(vin, ifmd, VNCSI_IFMD_REG);
 
 	vin_dbg(vin, "Set IFMD 0x%x\n", ifmd);
diff --git a/drivers/media/platform/rcar-vin/rcar-vin.h b/drivers/media/platform/rcar-vin/rcar-vin.h
index 76b034a444e8..b87ef559b820 100644
--- a/drivers/media/platform/rcar-vin/rcar-vin.h
+++ b/drivers/media/platform/rcar-vin/rcar-vin.h
@@ -44,6 +44,8 @@
 
 #define MSTP_WAIT_TIME 100
 
+#define RCAR_VIN_DES1_RESERVED		BIT(0)
+
 struct rvin_group;
 
 enum model_id {
@@ -206,6 +208,7 @@ struct rvin_info {
  * @chsel:		channel selection
  * @setup_wait:		wait queue used to setup VIN
  * @suspend:		suspend flag
+ * @chip_info:         chip information by each device
  */
 struct rvin_dev {
 	struct device *dev;
@@ -250,6 +253,7 @@ struct rvin_dev {
 	unsigned int chsel;
 	wait_queue_head_t setup_wait;
 	bool suspend;
+	u32 chip_info;
 };
 
 #define vin_to_source(vin)		((vin)->digital->subdev)
-- 
2.17.1

